File: apps/web/lib/correlation.ts
/**
 * Correlation ids for the web app (T0007, docs/26 §2).
 *
 * One id traces a page render across every boundary the render crosses:
 * the browser/user hits the web app (optionally carrying X-Correlation-ID),
 * the web app passes the id to the Go API and the scientific adapter via
 * the same header, and every service logs it. The id is also rendered on
 * the status page so a human can copy it and grep every log for it.
 *
 * Accepted shape is the shared one across the monorepo (Go internal/
 * observability, Python post_scientific_adapter.correlation): 8-64 chars
 * of letters, digits, '.', '_' or '-'. Anything else is rejected and a
 * fresh id is generated — an untrusted header must never reach a log.
 *
 * This module has no framework imports; it runs on the Node server side.
 */

import { randomUUID } from "node:crypto";

export const CORRELATION_HEADER = "X-Correlation-ID";

const CORRELATION_RE = /^[A-Za-z0-9][A-Za-z0-9._-]{7,63}$/;

/** Create a fresh correlation id (UUID v4 — 36 chars, within the shape). */
export function newCorrelationId(): string {
  return randomUUID();
}

/** Validate an incoming correlation id. */
export function isValidCorrelationId(id: string): boolean {
  return CORRELATION_RE.test(id);
}

/**
 * Resolve the correlation id for a request: honour a valid incoming header
 * value, otherwise create a fresh edge id.
 */
export function resolveCorrelationId(header: string | null | undefined): string {
  if (header !== null && header !== undefined && header !== "" && isValidCorrelationId(header)) {
    return header;
  }
  return newCorrelationId();
}

/** Fetch options carrying the correlation id (the id, never extra secrets). */
export function correlationHeaders(id: string): Record<string, string> {
  return { [CORRELATION_HEADER]: id };
}

File: services/scientific-adapter/src/post_scientific_adapter/correlation.py
"""Correlation ids for the scientific adapter (T0007, docs/26 §2).

One id traces a request across every service boundary: the Go API, the web
app and this adapter all speak the same ``X-Correlation-ID`` header and the
same id shape, so a caller (web app or API) can hand its id to the adapter
and every adapter log line carries it.

The accepted shape is the shared monorepo one (Go internal/observability,
web lib/correlation): 8-64 chars of letters, digits, '.', '_' or '-'.
Anything else is rejected and a fresh id is generated — an untrusted
header must never reach a log.

This module is stdlib-only, like the rest of the adapter scaffold.
"""

from __future__ import annotations

import re
import secrets

CORRELATION_HEADER = "X-Correlation-ID"

# Same shape as Go's internal/observability and the web app: first char
# alnum, then 7-63 chars of alnum / '.' / '_' / '-'.
_CORRELATION_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{7,63}$")


def new_correlation_id() -> str:
    """Create a fresh correlation id (32 hex chars, same shape as Go's)."""
    return secrets.token_hex(16)


def is_valid_correlation_id(value: str) -> bool:
    """Validate an incoming correlation id against the shared shape."""
    return bool(_CORRELATION_RE.match(value))


def resolve_correlation_id(header: str | None) -> str:
    """Honour a valid incoming header value, else create a fresh edge id."""
    if header and is_valid_correlation_id(header):
        return header
    return new_correlation_id()
